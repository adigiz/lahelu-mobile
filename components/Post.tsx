import MaterialCommunityIcons from '@expo/vector-icons/MaterialCommunityIcons';
import MaterialIcons from '@expo/vector-icons/MaterialIcons';
import { FlatList, Image, Text, TouchableOpacity, View } from "react-native";

import date from "@/utils/date";

import PostTag from "./PostTag";
import VideoPlayer from "./VideoPlayer";

interface PostProps {
  item: {
    type: string;
    title: string;
    user: {
      username: string;
      image_url: string;
    };
    file_url: string;
    uploaded_date: string;
    tags: string[];
    analytics: {
      upvote: number;
      total_comments: number;
    };
  };
  isVisible: boolean;
}

export default function Post({ item, isVisible }: PostProps) {
  const isVideo = item.type === "video"
  return (
    <View className="w-full bg-white py-[10]">
      <View className="flex flex-row items-center justify-between px-[10] pb-[10]">
        <View className="flex flex-row items-center">
          <View className="size-[30] overflow-hidden rounded-[30]">
            <Image
              source={{ uri: item.user.image_url }}
              className="size-[30]"
            />
          </View>
          <Text className="ml-[10] text-xs font-semibold">{item.user.username}</Text>
          <Text className="ml-[10] text-xs font-extralight">{date.timeAgo(item.uploaded_date)}</Text>
        </View>
        <TouchableOpacity>
          <MaterialCommunityIcons name="dots-horizontal" size={20} color="black" />
        </TouchableOpacity>
      </View>
      <View className="flex">
        <Text className="px-[10] text-lg font-semibold">{item.title}</Text>
        {isVideo ? <VideoPlayer uri={item.file_url} isPlaying={isVisible} /> : <Image source={{ uri: item.file_url }} className="aspect-square w-full" />}
      </View>
      <View className="flex flex-row p-[10]">
        <FlatList
          contentContainerStyle={{ gap: 5 }}
          data={item.tags}
          renderItem={({ item }) => <PostTag tag={item} />}
          keyExtractor={(item, index) => index.toString()}
          horizontal
          showsHorizontalScrollIndicator={false}
        />
      </View>
      <View className="flex flex-row justify-between px-[10]">
        <View className="flex flex-row gap-[8]">
          <View className="flex flex-row">
            <TouchableOpacity>
              <View className="flex flex-row items-center gap-[4] rounded-l-[5] border border-gray-300 p-[8]">
                <MaterialCommunityIcons name="arrow-up-bold-outline" size={20} color="black" />
                <Text>{item.analytics.upvote}</Text>
              </View>
            </TouchableOpacity>
            <TouchableOpacity>
              <View className="flex flex-row items-center rounded-r-[5] border-y border-r border-gray-300 p-[8]">
                <MaterialCommunityIcons name="arrow-down-bold-outline" size={20} color="black" />
              </View>
            </TouchableOpacity>
          </View>
          <TouchableOpacity>
            <View className="flex flex-row items-center gap-[4] rounded-[5] border border-gray-300 p-[8]">
              <MaterialIcons name="comment" size={20} color="black" />
              <Text>{item.analytics.total_comments}</Text>
            </View>
          </TouchableOpacity>
        </View>
        <TouchableOpacity>
          <View className="flex flex-row items-center gap-[4] rounded-[5] border border-gray-300 p-[8]">
            <MaterialCommunityIcons name="share" size={20} color="black" />
          </View>
        </TouchableOpacity>
      </View>
    </View>
  )
}