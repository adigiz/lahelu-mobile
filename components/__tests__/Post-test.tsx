import { render } from '@testing-library/react-native';
import React from 'react';

import Post from '@/components/Post';

jest.mock('@/utils/date', () => ({
  timeAgo: jest.fn(() => "5 Jam yang lalu"),
}));

describe('Post', () => {
  const mockItem = {
    type: 'image',
    title: 'Example Post Title',
    user: {
      username: 'john_doe',
      image_url: 'https://example.com/user_image.png',
    },
    file_url: 'https://example.com/image.png',
    uploaded_date: new Date().toISOString(),
    tags: ['tag1', 'tag2', 'tag3'],
    analytics: {
      upvote: 120,
      total_comments: 34,
    },
  };

  it('should matches the snapshot', () => {
    const { toJSON } = render(<Post item={mockItem} isVisible={true} />);
    
    expect(toJSON()).toMatchSnapshot();
  });
});
