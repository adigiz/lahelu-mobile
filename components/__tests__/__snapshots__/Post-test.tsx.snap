// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Post should matches the snapshot 1`] = `
<View
  className="w-full bg-white py-[10]"
>
  <View
    className="flex flex-row items-center justify-between px-[10] pb-[10]"
  >
    <View
      className="flex flex-row items-center"
    >
      <View
        className="size-[30] overflow-hidden rounded-[30]"
      >
        <Image
          className="size-[30]"
          source={
            {
              "uri": "https://example.com/user_image.png",
            }
          }
        />
      </View>
      <Text
        className="ml-[10] text-xs font-semibold"
      >
        john_doe
      </Text>
      <Text
        className="ml-[10] text-xs font-extralight"
      >
        5 Jam yang lalu
      </Text>
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={false}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
    >
      <Text />
    </View>
  </View>
  <View
    className="flex"
  >
    <Text
      className="px-[10] text-lg font-semibold"
    >
      Example Post Title
    </Text>
    <Image
      className="aspect-square w-full"
      source={
        {
          "uri": "https://example.com/image.png",
        }
      }
    />
  </View>
  <View
    className="flex flex-row p-[10]"
  >
    <RCTScrollView
      contentContainerStyle={
        {
          "gap": 5,
        }
      }
      data={
        [
          "tag1",
          "tag2",
          "tag3",
        ]
      }
      getItem={[Function]}
      getItemCount={[Function]}
      horizontal={true}
      keyExtractor={[Function]}
      onContentSizeChange={[Function]}
      onLayout={[Function]}
      onMomentumScrollBegin={[Function]}
      onMomentumScrollEnd={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      removeClippedSubviews={false}
      renderItem={[Function]}
      scrollEventThrottle={0.0001}
      showsHorizontalScrollIndicator={false}
      stickyHeaderIndices={[]}
      viewabilityConfigCallbackPairs={[]}
    >
      <View>
        <View
          onFocusCapture={[Function]}
          onLayout={[Function]}
          style={
            [
              {
                "flexDirection": "row",
              },
              null,
            ]
          }
        >
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={false}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "opacity": 1,
              }
            }
          >
            <View
              onLayout={[Function]}
              style={[]}
            >
              <Text>
                # tag1
              </Text>
            </View>
          </View>
        </View>
        <View
          onFocusCapture={[Function]}
          onLayout={[Function]}
          style={
            [
              {
                "flexDirection": "row",
              },
              null,
            ]
          }
        >
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={false}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "opacity": 1,
              }
            }
          >
            <View
              onLayout={[Function]}
              style={[]}
            >
              <Text>
                # tag2
              </Text>
            </View>
          </View>
        </View>
        <View
          onFocusCapture={[Function]}
          onLayout={[Function]}
          style={
            [
              {
                "flexDirection": "row",
              },
              null,
            ]
          }
        >
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={false}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "opacity": 1,
              }
            }
          >
            <View
              onLayout={[Function]}
              style={[]}
            >
              <Text>
                # tag3
              </Text>
            </View>
          </View>
        </View>
      </View>
    </RCTScrollView>
  </View>
  <View
    className="flex flex-row justify-between px-[10]"
  >
    <View
      className="flex flex-row gap-[8]"
    >
      <View
        className="flex flex-row"
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={false}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "opacity": 1,
            }
          }
        >
          <View
            className="flex flex-row items-center gap-[4] rounded-l-[5] border border-gray-300 p-[8]"
          >
            <Text />
            <Text>
              120
            </Text>
          </View>
        </View>
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={false}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "opacity": 1,
            }
          }
        >
          <View
            className="flex flex-row items-center rounded-r-[5] border-y border-r border-gray-300 p-[8]"
          >
            <Text />
          </View>
        </View>
      </View>
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={false}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "opacity": 1,
          }
        }
      >
        <View
          className="flex flex-row items-center gap-[4] rounded-[5] border border-gray-300 p-[8]"
        >
          <Text />
          <Text>
            34
          </Text>
        </View>
      </View>
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={false}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
    >
      <View
        className="flex flex-row items-center gap-[4] rounded-[5] border border-gray-300 p-[8]"
      >
        <Text />
      </View>
    </View>
  </View>
</View>
`;
